src/server/input_service.rs

```diff
@@ -28,14 +30,27 @@ impl super::service::Reset for StateCursor {
     }
 }
 
-#[derive(Default)]
 struct StatePos {
     cursor_pos: (i32, i32),
 }
 
+impl Default for StatePos {
+    fn default() -> Self {
+        Self {
+            cursor_pos: (INVALID_CURSOR_POS, INVALID_CURSOR_POS),
+        }
+    }
+}
+
 impl super::service::Reset for StatePos {
     fn reset(&mut self) {
-        self.cursor_pos = (0, 0);
+        self.cursor_pos = (INVALID_CURSOR_POS, INVALID_CURSOR_POS);
+    }
+}
+
+impl StatePos {
+    fn is_valid(&self) -> bool {
+        self.cursor_pos.0 != INVALID_CURSOR_POS && self.cursor_pos.1 != INVALID_CURSOR_POS
```

Comment:
I think only checking self.cursor_pos.0 != INVALID_CURSOR_POS is enough, and this function can be inline `#[inline]`